flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp

```diff
@@ -420,27 +420,41 @@ void ArrayCopyAnalysis::arrayMentions(
   loadMapSets.insert({load, visited});
 }
 
+static bool hasPointerAttribute(mlir::Type type) {
+  if (auto boxTy = type.dyn_cast<BoxType>())
+    type = boxTy.getEleTy();
+  return type.isa<fir::PointerType>();
+}
+
 /// Is there a conflict between the array value that was updated and to be
 /// stored to `st` and the set of arrays loaded (`reach`) and used to compute
 /// the updated value?
 static bool conflictOnLoad(llvm::ArrayRef<mlir::Operation *> reach,
                            ArrayMergeStoreOp st) {
   mlir::Value load;
-  auto addr = st.memref();
+  bool storeMemRefWasSeen = false;
+  mlir::Value addr = st.memref();
+  const bool storeHasPointerAddtribute = hasPointerAttribute(addr.getType());
   auto stEleTy = dyn_cast_ptrOrBoxEleTy(addr.getType());
   for (auto *op : reach)
     if (auto ld = mlir::dyn_cast<ArrayLoadOp>(op)) {
-      auto ldTy = ld.memref().getType();
-      if (auto boxTy = ldTy.dyn_cast<BoxType>())
-        ldTy = boxTy.getEleTy();
-      if (ldTy.isa<fir::PointerType>() && stEleTy == dyn_cast_ptrEleTy(ldTy))
-        return true;
+      mlir::Type ldTy = ld.memref().getType();
       if (ld.memref() == addr) {
         if (ld.getResult() != st.original())
           return true;
-        if (load)
+        if (storeMemRefWasSeen)
           return true;
-        load = ld;
+        storeMemRefWasSeen = true;
+      } else if (hasPointerAttribute(ldTy) &&
+                 stEleTy == dyn_cast_ptrOrBoxEleTy(ldTy)) {
```

Comment:
Since you've moved this to the bottom, we're already assured that the array_load being tested is not the one from the array_merge_store. The conflicting cases are if either the load or the store is a POINTER and the other is a POINTER or a TARGET and they have the same intrinsic/derived type. There doesn't seem to be a need for two different `else` cases.

Reply:
Good point, I have merged them.
Also, I am not entirely sure about the _the same intrinsic/derived type_ constraints given one can make complex component ref/derived type component ref, although I will keep it here for now since this logic was there already.

Reply:
I should have said "same TKR" to be more precise, but that's not the entire story either. A pointer can point into the "interior" of an object (as a projection).

Reply:
> So this part of the test can be improved.

Could you be more specific, do you want a change in this PR here ?

Reply:
No. I mean as a TODO.